test/e2e/util.go

```diff
@@ -0,0 +1,269 @@
+// and in the given namespace that curls the specified host and address.
+func buildCurlPod(name, namespace, host, address string, extraArgs ...string) *corev1.Pod {
+	curlArgs := []string{
+		"-s",
+		"-v",
+		"--header", "HOST:" + host,
+		"--retry", "300", "--retry-delay", "5", "--max-time", "2",
+	}
+	curlArgs = append(curlArgs, extraArgs...)
+	curlArgs = append(curlArgs, "http://"+address)
+	return &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:    "curl",
+					Image:   "openshift/origin-node",
+					Command: []string{"/bin/curl"},
+					Args:    curlArgs,
+				},
+			},
+			RestartPolicy: corev1.RestartPolicyNever,
+		},
+	}
+}
+
+// buildEchoService returns a service definition for an HTTP service.
+func buildEchoService(name, namespace string) *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeNodePort,
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "http",
+					Port:       int32(80),
+					Protocol:   corev1.ProtocolTCP,
+					TargetPort: intstr.FromInt(8080),
+				},
+			},
+			Selector: map[string]string{
+				"app": "echo",
+			},
+		},
+	}
+}
+
+func buildIngressRule(host string, path *networkingv1.HTTPIngressRuleValue) networkingv1.IngressRule {
+	return networkingv1.IngressRule{
+		Host: host,
+		IngressRuleValue: networkingv1.IngressRuleValue{
+			HTTP: path,
+		},
+	}
+}
+
+func buildIngressPath(path string, pathtype networkingv1.PathType, backend networkingv1.IngressBackend) networkingv1.HTTPIngressPath {
+	return networkingv1.HTTPIngressPath{
+		Path:     path,
+		PathType: &pathtype,
+		Backend:  backend,
+	}
+}
+
+func buildIngressBackend(svc *corev1.Service) networkingv1.IngressBackend {
+	return networkingv1.IngressBackend{
+		Service: &networkingv1.IngressServiceBackend{
+			Name: svc.Name,
+			Port: networkingv1.ServiceBackendPort{
+				Number: svc.Spec.Ports[0].Port,
+			},
+		},
+	}
+}
+
+type ingressbuilder struct {
+	name         types.NamespacedName
+	annotations  map[string]string
+	ingressclass string
+	rules        []networkingv1.IngressRule
+}
+
+func newIngressBuilder() *ingressbuilder {
+	return &ingressbuilder{
+		annotations:  make(map[string]string),
+		ingressclass: "alb",
+	}
+}
+
+func (b *ingressbuilder) withName(name types.NamespacedName) *ingressbuilder {
+	b.name = name
+	return b
+}
+
+func (b *ingressbuilder) withAnnotations(annotations map[string]string) *ingressbuilder {
+	b.annotations = annotations
+	return b
+}
+
+func (b *ingressbuilder) withIngressClass(class string) *ingressbuilder {
+	b.ingressclass = class
+	return b
+}
+
+func (b *ingressbuilder) withRules(rules []networkingv1.IngressRule) *ingressbuilder {
+	b.rules = rules
+	return b
+}
+
+func (b ingressbuilder) build() *networkingv1.Ingress {
```

Comment:
Using an builder is nice, but the `build()` doesn't have defaults for when the options are not set.

Reply:
Fixed.